File: 002. Add Two Numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* head = l1;
        ListNode* p = NULL;
        while (l1 != NULL && l2 != NULL){
            p = l1;
            l1->val += l2->val;
            l1 = l1->next;
            l2 = l2->next;
        }
        
        if (l2 != NULL){
            p->next = l2;
        }
        
        p = head;
        while( p->next != NULL){
            if ( p->val > 9){
                p->next->val += p->val / 10;
                p->val = p->val % 10;
            }
            p = p->next;
        }
        if (p->val > 9){
            p->next = new ListNode(p->val / 10);
            p->val = p->val % 10;
        }

        return head;
    }
};

File: 002. Add Two Numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        head = p = ListNode(0)
        
        while l1 or l2 or carry:
            var1 = var2 = 0
            if l1:
                var1 = l1.val
                l1 = l1.next
            if l2:
                var2 = l2.val
                l2 = l2.next
            carry,val = divmod(var1 + var2 + carry , 10)
            p.next = ListNode(val)
            p = p.next
            
        return head.next